### PR TITLE
feature: fetch e2ee chats and sync to local database

### PR DESCRIPTION
**Pull request by ice-kreios**
## Description
- New `fetchAndSyncConversations` provider for e2ee chat syncing
- Added persistent subscriptions support in NostrNotifier
- Refactored gift wrap and seal services with proper annotations

## Additional Notes
N/A

## Type of Change
- [ ] Bug fix
- [x] New feature
- [ ] Breaking change
- [x] Refactoring
- [ ] Documentation
- [ ] Chore

### CODE REVIEW

File: lib/app/features/chat/providers/fetch_conversation_provider.c.dart [hunk 1]
```diff
@@ -0,0 +1,68 @@
...
+import 'package:ion/app/services/nostr/ion_connect_gift_wrap_service.c.dart';
+import 'package:ion/app/services/nostr/ion_connect_seal_service.c.dart';
+import 'package:nostr_dart/nostr_dart.dart';
+import 'package:riverpod_annotation/riverpod_annotation.dart';
+
+part 'fetch_conversation_provider.c.g.dart';
+
+@riverpod
+Future<void> fetchAndSyncConversations(Ref ref) async {
+  final pubkey = ref.read(currentPubkeySelectorProvider);
```
Comment (ice-alcides) [inline lib/app/features/chat/providers/fetch_conversation_provider.c.dart:20]: replace please with `ref.watch` in this place

File: lib/app/features/chat/providers/fetch_conversation_provider.c.dart [hunk 2]
```diff
@@ -0,0 +1,68 @@
...
+
+  final requestMessage = RequestMessage()..addFilter(requestFilter);
+
+  final nostrEvents = ref.read(nostrNotifierProvider.notifier).requestEvents(
+        requestMessage,
+        actionSource: const ActionSourceCurrentUserChat(),
+        keepSubscription: true,
+      );
+
+  final currentUserSigner = await ref.read(currentUserNostrEventSignerProvider.future);
```
Comment (ice-alcides) [inline lib/app/features/chat/providers/fetch_conversation_provider.c.dart:49]: ref.watch

File: lib/app/features/chat/providers/fetch_conversation_provider.c.dart [hunk 3]
```diff
@@ -0,0 +1,68 @@
...
+import 'package:nostr_dart/nostr_dart.dart';
+import 'package:riverpod_annotation/riverpod_annotation.dart';
+
+part 'fetch_conversation_provider.c.g.dart';
+
+@riverpod
+Future<void> fetchAndSyncConversations(Ref ref) async {
+  final pubkey = ref.read(currentPubkeySelectorProvider);
+  if (pubkey == null) {
+    return;
```
Comment (ice-damocles) [inline lib/app/features/chat/providers/fetch_conversation_provider.c.dart:22]: `throw EventSignerNotFoundException()`

File: lib/app/features/chat/providers/fetch_conversation_provider.c.dart [hunk 4]
```diff
@@ -0,0 +1,68 @@
...
+  final requestMessage = RequestMessage()..addFilter(requestFilter);
+
+  final nostrEvents = ref.read(nostrNotifierProvider.notifier).requestEvents(
+        requestMessage,
+        actionSource: const ActionSourceCurrentUserChat(),
+        keepSubscription: true,
+      );
+
+  final currentUserSigner = await ref.read(currentUserNostrEventSignerProvider.future);
+  if (currentUserSigner == null) return;
```
Comment (ice-damocles) [inline lib/app/features/chat/providers/fetch_conversation_provider.c.dart:50]: `throw EventSignerNotFoundException()`

File: lib/app/features/chat/providers/fetch_conversation_provider.c.dart [hunk 5]
```diff
@@ -0,0 +1,68 @@
...
+          currentUserSigner,
+        );
+
+    final unwrappedSeal = await ref.read(ionConnectSealServiceProvider).decodeSeal(
+          unwrappedGift,
+          currentUserSigner,
+          pubkey,
+        );
+    await dbProvider.insertEventMessage(unwrappedSeal);
+  }
```
Comment (ice-endymion) [inline lib/app/features/chat/providers/fetch_conversation_provider.c.dart:67]: Maybe wrap it with try catch to log errors, and we can use Future.wait here.

**Review by ice-endymion — CHANGES_REQUESTED** [review]

File: lib/app/features/chat/providers/fetch_conversation_provider.c.dart [hunk 6]
```diff
@@ -0,0 +1,68 @@
...
+import 'package:ion/app/services/database/ion_database.c.dart';
+import 'package:ion/app/services/nostr/ion_connect_gift_wrap_service.c.dart';
+import 'package:ion/app/services/nostr/ion_connect_seal_service.c.dart';
+import 'package:nostr_dart/nostr_dart.dart';
+import 'package:riverpod_annotation/riverpod_annotation.dart';
+
+part 'fetch_conversation_provider.c.g.dart';
+
+@riverpod
+Future<void> fetchAndSyncConversations(Ref ref) async {
```
Comment (ice-tychon) [inline lib/app/features/chat/providers/fetch_conversation_provider.c.dart:19]: Don't do DB changes during the build of the provider.
https://riverpod.dev/docs/essentials/do_dont#dont-perform-side-effects-during-the-initialization-of-a-provider

File: lib/app/features/chat/providers/fetch_conversation_provider.c.dart (hunk 1 above)
Comment (ice-kreios) [reply lib/app/features/chat/providers/fetch_conversation_provider.c.dart]: fixed

File: lib/app/features/chat/providers/fetch_conversation_provider.c.dart (hunk 2 above)
Comment (ice-kreios) [reply lib/app/features/chat/providers/fetch_conversation_provider.c.dart]: fixed

File: lib/app/features/chat/providers/fetch_conversation_provider.c.dart (hunk 5 above)
[the remark this answers is not in the document]
Comment (ice-kreios) [reply lib/app/features/chat/providers/fetch_conversation_provider.c.dart:83]: fixed

Comment (ice-kreios) [reply lib/app/features/chat/providers/fetch_conversation_provider.c.dart]: fixed

File: lib/app/features/chat/providers/fetch_conversation_provider.c.dart (hunk 6 above)
Comment (ice-kreios) [reply lib/app/features/chat/providers/fetch_conversation_provider.c.dart]: fixed

File: lib/app/features/chat/providers/fetch_conversation_provider.c.dart (hunk 3 above)
Comment (ice-kreios) [reply lib/app/features/chat/providers/fetch_conversation_provider.c.dart:23]: fixed

File: lib/app/features/chat/providers/fetch_conversation_provider.c.dart (hunk 4 above)
Comment (ice-kreios) [reply lib/app/features/chat/providers/fetch_conversation_provider.c.dart]: fixed

**Review by ice-tychon — APPROVED** [review]

**Review by ice-endymion — DISMISSED** [review]

File: lib/app/features/chat/providers/fetch_conversation_provider.c.dart [hunk 7]
```diff
@@ -0,0 +1,82 @@
...
+    return;
+  }
+
+  Future<void> fetchAndSync() async {
+    final pubkey = ref.watch(currentPubkeySelectorProvider);
+    if (pubkey == null) {
+      throw UserMasterPubkeyNotFoundException();
+    }
+
+    final currentUserSigner = await ref.watch(currentUserNostrEventSignerProvider.future);
```
Comment (ice-tychon) [inline lib/app/features/chat/providers/fetch_conversation_provider.c.dart:31]: ref.watch

**Review by ice-tychon — APPROVED** [review]

File: lib/app/features/chat/providers/fetch_conversation_provider.c.dart (hunk 7 above)
Comment (ice-kreios) [reply lib/app/features/chat/providers/fetch_conversation_provider.c.dart]: fixed

**Review by ice-brontes — APPROVED** [review]

**Review by ice-alcides — APPROVED** [review]

**Pull request merged**